✨ [#86] 경매 상품 목록 검색 api 로직 변경

puclpu opened the pull request:
## #️⃣연관된 이슈
- 이슈 번호: #86

## 작업 내용
- 경매 상품 목록 검색 api 로직 변경했습니다.
   - 전체 검색, 필터링 검색, 내가 등록한 상품 목록 검색이 가능합니다. 
   - 경매 시작 시각, 경매가 시작 될 상품 상태를 변경(UPCOMING >> ONGOING)됩니다.
   - 경매 마감 시각, 경매가 마감 될 상품 상태를 변경(ONGOING >> ENDED)됩니다.
## 체크리스트
- [x] 코드가 제대로 동작하는지 확인했습니다.

puclpu committed:
✨ [#86] 경매 상품 목록 검색 api 로직 변경

- ProductStatus 추가
- 전체 목록과 필터링 목록 조회 분리
- Scheduler 를 사용해 일정 시간 자동으로 상태 변경

puclpu committed:
Merge remote-tracking branch 'origin/dev' into feature/#86

# Conflicts:
#	goodsending/src/main/java/com/goodsending/product/controller/ProductController.java
#	goodsending/src/main/java/com/goodsending/product/repository/ProductCustomRepository.java
#	goodsending/src/main/java/com/goodsending/product/repository/ProductCustomRepositoryImpl.java
#	goodsending/src/main/java/com/goodsending/product/service/ProductService.java
#	goodsending/src/main/java/com/goodsending/product/service/ProductServiceImpl.java

puclpu committed:
♻️ [#86] 경매 상품 목록 조회 통합

- 전체 상품 목록 조회, 필터링 상품 목록 조회, 내가 등록한 상품 목록 조회를 통합

puclpu committed:
Merge branch 'dev' into feature/#86

je-pa commented on goodsending/src/main/java/com/goodsending/product/repository/ProductCustomRepositoryImpl.java:
```diff
@@ -30,207 +29,83 @@ public class ProductCustomRepositoryImpl implements ProductCustomRepository {
+          .limit(pageable.getPageSize()+1);
+    }
+
+    List<ProductSummaryDto> productSummaryDtoList = query.fetch();
```
이때 쿼리 select문이 나가서 결과적으로 해당 메서드에서는 한번의 select 쿼리만 실행되는거죠?

puclpu replied:
네 맞습니다!

je-pa commented on goodsending/src/main/java/com/goodsending/product/repository/ProductCustomRepositoryImpl.java:
```diff
@@ -30,207 +29,83 @@ public class ProductCustomRepositoryImpl implements ProductCustomRepository {
+      query.where(productImageEq().and(cursorBuilder).and(keywordBuilder).and(closedExpression))
+          .orderBy(product.startDateTime.asc(), product.id.asc())
+          .limit(pageable.getPageSize()+1);
+    }
```
중복되는 코드가 조금 있는데 where 절에 null 들어가면 where 조건에서 자동으로 해당 조건문은 건너 뛰는 걸로 알아요!
고거 이용해서 추후에 리펙토링 해보셔도 좋을 것 같아요~

![image](https://github.com/user-attachments/assets/472934d2-6354-4122-b37f-5babc80613f3)

참고링크: https://jojoldu.tistory.com/528

puclpu replied:
중복코드 줄이는 방향으로 리팩토링 했습니다!

puclpu committed:
🐛 [#86] QProductSummaryDto status 추가

je-pa commented on goodsending/src/main/java/com/goodsending/product/repository/ProductCustomRepositoryImpl.java:
```diff
@@ -243,9 +118,36 @@ public List<Product> findTop5ByStartDateTimeAfterOrderByLikeCountDesc(
+    NumberExpression<Integer> statusRank = new CaseBuilder()
+        .when(product.status.eq(ProductStatus.ONGOING)).then(ProductStatus.ONGOING.getRank())
+        .when(product.status.eq(ProductStatus.UPCOMING)).then(ProductStatus.UPCOMING.getRank())
+        .when(product.status.eq(ProductStatus.ENDED)).then(ProductStatus.ENDED.getRank())
+        .otherwise(4);
+    return statusRank;
+  }
```
rank해서 정렬하는 방법 배워갑니다~

puclpu committed:
♻️ [#86] openProduct, closedProduct 타입 변경 (String > boolean)

je-pa commented on goodsending/src/main/java/com/goodsending/product/service/ProductServiceImpl.java:
```diff
@@ -249,20 +250,20 @@ public void deleteProduct(Long productId, Long memberId, LocalDateTime now) {
+    List<Product> products = productRepository.findAllByStatusAndStartDateTime(status, startDateTime);
+    for (Product product : products) {
+      switch (status) {
+        case UPCOMING:
+          product.setStatus(ProductStatus.ONGOING);
+          break;
+        case ONGOING:
+          product.setStatus(ProductStatus.ENDED);
```
UPCOMING이 진행될 예정이고
ONGOING이 진행중이고
ENDED가 끝난 인거죠?

puclpu replied:
네 맞습니다!

je-pa commented on goodsending/src/main/java/com/goodsending/product/service/ProductServiceImpl.java:
```diff
@@ -249,20 +250,20 @@ public void deleteProduct(Long productId, Long memberId, LocalDateTime now) {
-    return myProductSummaryDtoList;
+  @Transactional
+  public void updateProductStatus(ProductStatus status) {
+    LocalDateTime startDateTime = LocalDateTime.now().withMinute(0).withSecond(0).withNano(0);
+    List<Product> products = productRepository.findAllByStatusAndStartDateTime(status, startDateTime);
```
updateOnGoingProduct() 일때도 startDateTime으로 들고오는게 맞을까요?

LocalDateTime 을 service에서 받기전에 만들어서 파라미터로 받는게 좀더 메서드가 명확해지고 나중에 테스트할때도 편할 것 같아요.

puclpu replied:
LocalDateTime 생성 위치 ProductScheduler로 변경해주었습니다

je-pa reviewed (commented):
고생하셨습니다~

puclpu committed:
♻️ [#86] startDateTime 생성 위치 변경

- ProductService > ProductScheduler

je-pa commented on goodsending/src/main/java/com/goodsending/product/scheduler/ProductScheduler.java:
```diff
@@ -0,0 +1,29 @@
+
+  private final ProductService productService;
+
+  @Scheduled(cron = "0 0 12,18 * * *") // 매일 12시, 18시
+  public void updateUpComingProduct() {
+    log.info("경매 진행 상태 전환 ");
+    productService.updateProductStatus(ProductStatus.UPCOMING);
+  }
+
+  @Scheduled(cron = "59 18 14,20 * * *") // 매일 14시, 20시 59분 59초
+  public void updateOnGoingProduct() {
+    log.info("경매 종료 상태 전환");
+    productService.updateProductStatus(ProductStatus.ONGOING);
+  }
```
PR 설명에 해당 내용도 추가해주시면 좋을 것 같아요!

puclpu replied:
추가했습니다!

puclpu committed:
♻️ [#86] 검색 조건 Dto 생성

- ProductSearchCondition 생성
- 필터링 검색 조건 BooleanBuilder 생성하여 쿼리 메서드와 조건 생성 메서드를 분리

puclpu committed:
💡 [#86] ProductController, ProductSearchCondition 주석 수정

je-pa commented on goodsending/src/main/java/com/goodsending/product/repository/ProductCustomRepositoryImpl.java:
```diff
@@ -30,222 +30,140 @@ public class ProductCustomRepositoryImpl implements ProductCustomRepository {
+    BooleanBuilder cursorBuilder = new BooleanBuilder();
+    if (cursorStatus != null && cursorStartDateTime != null && cursorId != null) {
+      cursorBuilder.and(
+          statusRank.gt(cursorStatus.getRank())
+              .or(statusRank.eq(cursorStatus.getRank()).and(product.startDateTime.gt(cursorStartDateTime)))
+              .or(statusRank.gt(cursorStatus.getRank()).and(product.startDateTime.eq(cursorStartDateTime)).and(product.id.gt(cursorId)))
```
이거 statusRank.gt(cursorStatus.getRank())에
statusRank.gt(cursorStatus.getRank()).and(product.startDateTime.eq(cursorStartDateTime)).and(product.id.gt(cursorId))가
포함되지 않나요?

혹시 제가 잘 이해를 못한거라면 cursorBuilder 메서드 설명좀 부탁드려도 될까요!

puclpu replied:
eq여야 하는데 잘못 기입했네요! 감사합니다🥲 수정해서 push했습니다!

je-pa reviewed (commented):
고생하셨습니다~

puclpu committed:
🐛 [#86] cursorBuilder 조건 버그 수정

- statusRank.gt >> statusRank.eq
